src/protocol/sort/generate_sort_permutations.rs

```diff
@@ -0,0 +1,131 @@
+use crate::{
+    error::BoxError,
+    ff::Field,
+    protocol::{
+        context::ProtocolContext,
+        modulus_conversion::convert_shares::convert_all_shares,
+        sort::bit_permutations::BitPermutations,
+        sort::SortStep::{ApplyInv, BitPermutation, ComposeStep},
+    },
+    secret_sharing::Replicated,
+};
+
+use super::{compose::Compose, secureapplyinv::SecureApplyInv};
+use embed_doc_image::embed_doc_image;
+
+/// This is an implementation of `GenPerm` (Algorithm 6) described in:
+/// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>.
+#[derive(Debug)]
+pub struct GenerateSortPermutations<'a> {
+    input: &'a [u64],
+}
+
+impl<'a> GenerateSortPermutations<'a> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a [u64]) -> GenerateSortPermutations {
+        Self { input }
+    }
+
+    #[allow(dead_code)]
+    #[embed_doc_image("semi_honest_sort", "images/sort/semi-honest-sort.png")]
+    /// This protocol generates permutation of a stable sort for the given inputs.
+    /// ![Generate sort permutations steps][semi_honest_sort]
+    /// Steps
+    /// 1. Obtain bit-wise shares in Field by calling modulus conversion on the input match keys.
+    /// 2. Obtain bit permutation to sort 0th input bit. This is also 0th bit composition (i.e. sigma)
+    /// 3. For 1st until n bits, following steps are repeated
+    /// 3i.  Apply inverse of i-1th bit permutation on ith input bits
+    /// 3ii. Obtain bit permutation to sort ith input bit
+    /// 3iii.Compose ith bit permutation on i-1th composition. We have now obtained ith composition
+    /// 4. Return nth composition as the permutation to sort the given inputs
+    pub async fn execute<F: Field>(
+        &self,
+        ctx: ProtocolContext<'_, F>,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let mut bits = convert_all_shares(&ctx, self.input).await?;
+        let mut sigma = BitPermutations::new(&bits[0])
```

Comment:
I see. You wanted the data in this format which is why you did that whole "transpose" thing at the end of the `convert_all_shares` function.

I'd like to avoid all of that copying bytes around. Is there a way that we can just keep things in the original format?

i.e.
```
      | bit n | bit n-1 | bit n-2 | ... | bit 3 | bit 2 | bit 1 |
row 1 |       |         |         | ... |       |       |       | 
row 2 |       |         |         | ... |       |       |       | 
 ...  |       |         |         | ... |       |       |       | 
row m |       |         |         | ... |       |       |       | 
```

You could get the shares of the nth bit by doing:
```
foo.iter().map(|bit_shares| { bit_shares[n] })
```
I feel like there should be some way to make that an "iterator" that just loops over things and doesn't need to deep-copy them. I think if you do not include a "collect()" then it is a lazy iterator?

Reply:
I don't think the variable name "sigma" is adding anything to comprehensibility here. Perhaps a name like "bit_zero_permutation" would be better.